examples/vision/semantic_image_clustering.py

```diff
@@ -0,0 +1,576 @@
+(SCAN)](https://arxiv.org/abs/2005.12320) algorithm, by Wouter Van Gansbeke et. al, on the
+[CIFAR-10](https://www.cs.toronto.edu/~kriz/cifar.html) images dataset. The algorithm,
+consists of two phases:
+1. Self-supervised visual representation learning of images, in which we use the
+[simCLR](https://arxiv.org/abs/2002.05709?ref=hackernoon.com) technique.
+2. Clustering of the learnt visual representation vectors that maximizes the agreement
+between the cluster assignments of the neighbours.
+
+The example requires [TensorFlow Addons](https://www.tensorflow.org/addons), 
+which you can install using the following command:
+
+```python
+pip install tensorflow-addons
+```
+"""
+"""
+## Setup
+"""
+
+from collections import defaultdict
+import random
+import numpy as np
+import tensorflow as tf
+import tensorflow_addons as tfa
+from tensorflow import keras
+from tensorflow.keras import layers
+import matplotlib.pyplot as plt
+from tqdm import tqdm
+
+"""
+## Prepare the data
+"""
+
+num_classes = 10
+input_shape = (32, 32, 3)
+
+(x_train, y_train), (x_test, y_test) = keras.datasets.cifar10.load_data()
+x_data = np.concatenate([x_train, x_test])
+y_data = np.concatenate([y_train, y_test])
+
+print("x_data shape:", x_data.shape, " - y_data shape:", y_data.shape)
+
+class_labels = [
+    "airplane",
+    "automobile",
+    "bird",
+    "cat",
+    "deer",
+    "dog",
+    "frog",
+    "horse",
+    "ship",
+    "truck",
+]
+
+"""
+## Define hyperparameters
+"""
+
+target_size = 32  # Resize the input images.
+representation_dim = 512  # The dimensions of the features vector.
+projection_units = 128  # The projection head of the representation learner.
+num_augumentations = 2  # Number of augmented images to generate for each input.
+num_clusters = 20  # Number of clusters.
+k_neighbours = 5  # Number of neighbours to consider during cluster learning.
+tune_encoder_during_clustering = False  # Freeze the encoder in the cluster learning.
+
+"""
+## Implement data preprocessing
+
+The data preprocessing resizes the input images to the desired `target_size`, and applies
+feature-wise normalization. Note that, when using `keras.applications.ResNet50V2` as the
+visual encoder, resizing the input images towards 255 x 255 would lead to better results,
+but would require longer time to train.
+"""
+
+data_preprocessing = keras.Sequential(
+    [
+        layers.experimental.preprocessing.Resizing(target_size, target_size),
+        layers.experimental.preprocessing.Normalization(),
+    ]
+)
+# Compute the mean and the variance from the data for normalization.
+data_preprocessing.layers[-1].adapt(x_data)
+
+"""
+## Data augmentation
+Unlink simCLR, which randomly picks a single data augmentation function to apply to an input
+image, we apply a set of data augmentation functions randomly to the input image.
+You can experiment with other [image augmentation functinoality](https://www.tensorflow.org/tutorials/images/data_augmentation).
```

Comment:
Please use a grammatical linter:

```suggestion
(You can experiment with other image augmentation techniques by following the [data augmentation tutorial](https://www.tensorflow.org/tutorials/images/data_augmentation).)
```